src/ooni/templates.cpp

```diff
@@ -0,0 +1,127 @@
+                   Entry query_entry;
+                   query_entry["resolver_hostname"] = resolver_hostname;
+                   query_entry["resolver_port"] = resolver_port;
+                   query_entry["failure"] = nullptr;
+                   query_entry["answers"] = Entry::array();
+                   if (query_type == dns::QueryTypeId::A) {
+                       query_entry["query_type"] = "A";
+                       query_entry["hostname"] = query_name;
+                   }
+                   if (!error) {
+                       for (auto answer : message.answers) {
+                           if (query_type == dns::QueryTypeId::A) {
+                               query_entry["answers"].push_back(
+                                   {{"ttl", answer.ttl},
+                                    {"ipv4", answer.ipv4},
+                                    {"answer_type", "A"}});
+                           }
+                       }
+                   } else {
+                       query_entry["failure"] = error.as_ooni_error();
+                   }
+                   // TODO add support for bytes received
+                   // query_entry["bytes"] = response.get_bytes();
+                   (*entry)["queries"].push_back(query_entry);
+                   logger->debug("dns_test: callbacking");
+                   cb(error, message);
+                   logger->debug("dns_test: callback called");
+               },
+               options, reactor);
+}
+
+void http_request(Var<Entry> entry, Settings settings, http::Headers headers,
+                  std::string body, Callback<Error, Var<http::Response>> cb,
+                  Var<Reactor> reactor, Var<Logger> logger) {
+
+    (*entry)["agent"] = "agent";
+    (*entry)["socksproxy"] = nullptr;
+
+    if (settings.find("http/method") == settings.end()) {
+        settings["http/method"] = "GET";
+    }
+
+    http::request(
+        settings, headers, body,
+        [=](Error error, Var<http::Response> response) {
+            Entry rr;
+            rr["request"]["headers"] =
+                std::map<std::string, std::string>(headers);
+            rr["request"]["body"] = body;
+            rr["request"]["url"] = settings.at("http/url").c_str();
+            rr["request"]["method"] = settings.at("http/method").c_str();
+
+            // XXX we should probably update OONI data format to remove this.
+            rr["method"] = settings.at("http/method").c_str();
+
+            if (!error) {
+                rr["response"]["headers"] =
+                    std::map<std::string, std::string>(response->headers);
+                rr["response"]["body"] = response->body;
+                rr["response"]["response_line"] = response->response_line;
+                rr["response"]["code"] = response->status_code;
+                rr["failure"] = nullptr;
+            } else {
+                rr["failure"] = error.as_ooni_error();
```

Comment:
Here `failure` was "uknown_error measurement_kit error" and the error code was also passed. Since now we have `as_ooni_error()` I refactored this only to set `failure`